### PR TITLE
5227 - Clarify process for running e2e tests locally

### PR DESCRIPTION
**Pull request by SCdF**
Added an e2e script that Works For Me™. We should probably consider
formalising it in Grunt.

Also refactored some e2e code that was redundant.

medic/medic-webapp#5227

### CODE REVIEW

File: tests/service-manager.js [hunk 1]
```diff
@@ -7,6 +7,7 @@ const processes = [];
 let stoppingAll = false;
 
 const stopAll = (code) => {
+  console.log(`[e2e] Sending kill signals to all servers due to code ${code}`);
```
Comment (SCdF) [inline tests/service-manager.js]: I moved this, because otherwise it's really not clear that stopAll is being fired

**Comment by kennsippell** [comment]
Seems easy to make this a grunt task initially -- rather than patching it later. `e2e.sh` is basically a grunt task waiting to happen + you can just make `e2e-servers.js` a grunt `exec` item in one line by having it execute `node scripts/e2e-servers.js`.

**Comment by SCdF** [comment]
I took a quick stab at this, but a direct copy of the script into grunt fails many tests, which passes when using the bash script.

```diff
index 957c7fc16..554b19435 100644
--- a/Gruntfile.js
+++ b/Gruntfile.js
@@ -447,6 +447,9 @@ module.exports = function(grunt) {
           )
           .join(' && '),
       },
+      'e2e-servers': {
+        cmd: 'node ./scripts/e2e-servers.js &'
+      },
       bundlesize: {
         cmd: 'node ./node_modules/bundlesize/index.js',
       },
@@ -961,11 +964,19 @@ module.exports = function(grunt) {

   // Test tasks
   grunt.registerTask('e2e', 'Deploy app for testing and run e2e tests', [
+    // 'exec:start-webdriver',
+    'exec:reset-test-databases',
+    // 'build-dev',
+    'couch-push:test',
+    'exec:e2e-servers',
+    'protractor:e2e-tests',
+  ]);
```

I'm not going to have time to look at this in more detail for awhile, so if anyone else wants to take a stab go for it, otherwise I'll pick it up in a couple of weeks maybe.

**Comment by dianabarsan** [comment]
@SCdF 
Yea, I'm aware API and Sentinel keep on running after the tests are done.
I didn't care very much about it as this was a quickly improvised solution to have tests running locally. `ServiceManager` theoretically does kill it's spawned processes when it's own process ends but that doesn't happen.
Even as it is, the test suite is re-runnable - when we delete the test database before running the tests, that effectively kills the previously running API & Sentinel, freeing the ports for the current execution. 

I might have a look at this if time permits next week.

**Comment by kennsippell** [comment]
What do you think of this?   I updated `service-manager` to use `child_process.fork` with `detached: false` to ensure the services are linked to the lifecycle of the parent process.  One test is failing for me locallly still - just pushing to see if it passes Travis.

**Comment by garethbowen** [comment]
@SCdF What do you think of Kenn's changes?

**Comment by SCdF** [comment]
@garethbowen / @kennsippell sorry it took me so long to reply to this.

I think the changes are good, though do you think we can also have the webdriver run automatically as well? Unless there is a reason we want to do that manually (imo it would be nice to have the one command that boots and then kills everything in one neat package)

**Comment by kennsippell** [comment]
@SCdF Done

**Pull request merged**